Dalamud/Interface/Internal/ReShadeHandling/ReShadeAddonInterface.cs

```diff
@@ -0,0 +1,176 @@
+            AddonEvent.DestroySwapChain,
+            this.destroySwapChainDelegate = new((ref ApiObject rt) => this.DestroySwapChain?.Invoke(ref rt)));
+    }
+
+    /// <summary>Finalizes an instance of the <see cref="ReShadeAddonInterface"/> class.</summary>
+    ~ReShadeAddonInterface() => this.ReleaseUnmanagedResources();
+
+    /// <summary>Delegate for <see cref="ReShadeAddonInterface.AddonEvent.ReShadeOverlay"/>.</summary>
+    /// <param name="effectRuntime">Reference to the ReShade runtime.</param>
+    public delegate void ReShadeOverlayDelegate(ref ApiObject effectRuntime);
+
+    /// <summary>Delegate for <see cref="ReShadeAddonInterface.AddonEvent.InitSwapChain"/>.</summary>
+    /// <param name="swapChain">Reference to the ReShade SwapChain wrapper.</param>
+    public delegate void ReShadeInitSwapChain(ref ApiObject swapChain);
+
+    /// <summary>Delegate for <see cref="ReShadeAddonInterface.AddonEvent.DestroySwapChain"/>.</summary>
+    /// <param name="swapChain">Reference to the ReShade SwapChain wrapper.</param>
+    public delegate void ReShadeDestroySwapChain(ref ApiObject swapChain);
+
+    private delegate BOOL GetModuleHandleExWDelegate(uint dwFlags, ushort* lpModuleName, HMODULE* phModule);
+
+    /// <summary>Called on <see cref="ReShadeAddonInterface.AddonEvent.ReShadeOverlay"/>.</summary>
+    public event ReShadeOverlayDelegate? ReShadeOverlay;
+
+    /// <summary>Called on <see cref="ReShadeAddonInterface.AddonEvent.InitSwapChain"/>.</summary>
+    public event ReShadeInitSwapChain? InitSwapChain;
+
+    /// <summary>Called on <see cref="ReShadeAddonInterface.AddonEvent.DestroySwapChain"/>.</summary>
+    public event ReShadeDestroySwapChain? DestroySwapChain;
+
+    /// <summary>Registers Dalamud as a ReShade addon.</summary>
+    /// <param name="r">Initialized interface.</param>
+    /// <returns><c>true</c> on success.</returns>
+    public static bool TryRegisterAddon([NotNullWhen(true)] out ReShadeAddonInterface? r)
+    {
+        try
+        {
+            r = Exports.ReShadeRegisterAddon is null ? null : new();
+            return r is not null;
+        }
+        catch
+        {
+            r = null;
+            return false;
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        this.ReleaseUnmanagedResources();
+        GC.SuppressFinalize(this);
+    }
+
+    private void ReleaseUnmanagedResources()
+    {
+        Exports.ReShadeUnregisterEvent(AddonEvent.InitSwapChain, this.initSwapChainDelegate);
+        Exports.ReShadeUnregisterEvent(AddonEvent.DestroySwapChain, this.destroySwapChainDelegate);
+        Exports.ReShadeUnregisterEvent(AddonEvent.ReShadeOverlay, this.reShadeOverlayDelegate);
+        Exports.ReShadeUnregisterAddon(this.hDalamudModule);
+        this.addonModuleResolverHook.Disable();
+        this.addonModuleResolverHook.Dispose();
+    }
+
+    private BOOL GetModuleHandleExWDetour(uint dwFlags, ushort* lpModuleName, HMODULE* phModule)
```

Comment:
It's a shame that this is necessary, but I guess there's no way around it for now. We should remove this as soon as possible when crosire releases a version with the HMODULE variants.

Reply:
I'll add a TODO about it.